File: solutions/haskell/complex-numbers/1/src/ComplexNumbers.hs
module ComplexNumbers
  ( Complex,
    conjugate,
    abs,
    exp,
    real,
    imaginary,
    mul,
    add,
    sub,
    div,
    complex,
  )
where

import Prelude hiding (abs, div, exp)

-- Data definition -------------------------------------------------------------
data Complex a = Complex a a deriving (Eq, Show)

complex :: (a, a) -> Complex a
complex (x, y) = Complex x y

-- unary operators -------------------------------------------------------------
conjugate :: (Num a) => Complex a -> Complex a
conjugate (Complex x y) = Complex x (-y)

abs :: (Floating a) => Complex a -> a
abs (Complex x y) = sqrt (x * x + y * y)

real :: (Num a) => Complex a -> a
real (Complex x y) = x

imaginary :: (Num a) => Complex a -> a
imaginary (Complex x y) = y

exp :: (Floating a) => Complex a -> Complex a
exp (Complex x y) = Complex (2.718 ** x * (cos y)) (2.718 ** x * (sin y))

-- binary operators ------------------------------------------------------------
mul :: (Num a) => Complex a -> Complex a -> Complex a
mul (Complex x y) (Complex z w) = Complex (x * z - y * w) (x * w + y * z)

add :: (Num a) => Complex a -> Complex a -> Complex a
add (Complex x y) (Complex z w) = Complex (x + z) (y + w)

sub :: (Num a) => Complex a -> Complex a -> Complex a
sub (Complex x y) (Complex z w) = Complex (x - z) (y - w)

div :: (Fractional a) => Complex a -> Complex a -> Complex a
div (Complex x y) (Complex z w) = Complex ((x * z + y * w) / (z * z + w * w)) ((y * z - x * w) / (z * z + w * w))

File: solutions/typescript/complex-numbers/1/complex-numbers.ts
export class ComplexNumber {
  public _real: number;
  public _imag: number;
  constructor(real: number, imaginary: number) {
    this._real = real;
    this._imag = imaginary;
  }

  public get real(): number {
    return this._real
  }

  public get imag(): number {
    return this._imag
  }

  public add(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this._real + other._real,
      this._imag + other._imag
    )
  }

  public sub(other: ComplexNumber): ComplexNumber {
    return new ComplexNumber(
      this._real - other._real,
      this._imag - other._imag
    )
  }

  public div(other: ComplexNumber): ComplexNumber {
    let a = this._real;
    let b = this._imag;
    let c = other._real;
    let d = other._imag;
    return new ComplexNumber((a*c+b*d)/(c**2+d**2), (b*c-a*d)/(c**2+d**2))
  }

  public mul(other: ComplexNumber): ComplexNumber {
    let a = this._real;
    let b = this._imag;
    let c = other._real;
    let d = other._imag;
    return new ComplexNumber(a*c-b*d, b*c+a*d)
  }

  public get abs(): number {
    return Math.sqrt(this.real**2+this.imag**2)
  }

  public get conj(): ComplexNumber {
    const a = this._real
    const b = this._imag === 0 ? 0 : this._imag * (-1)
    return new ComplexNumber(a, b)
  }

  public get exp(): ComplexNumber {
    let r = Math.exp(this._real)
    return new ComplexNumber(r * Math.cos(this._imag), r * Math.sin(this._imag))
  }
}
